Merge GitLab merge requests when shipping

manuelprinz opened the pull request:
Hi, this is my attempt to fix #1092. It is my first real contact with Go (except from some toying around), so some of it might look weird or not be best practice. I appreciate any feedback!

I had some issues with getting the test to run. There seemed to be an issue with some commit messages. I included a "fix" in this branch. If it is just on my box, I'm happy to revert it.

I used the GitHub driver as an example. The inner working is pretty much the same, so there is quite a bit of duplication, also in the tests. I had to shut up the linter, and am fully aware that this is not the right solution, but I did not know any better way for the time being. (As I said, Go beginner.) I guess that having something like a template method might be a solution, but some guidance would be appreciated.

It also worked when testing manually. I have not tested it with [all the merge strategies in GitLab](https://docs.gitlab.com/ee/user/project/merge_requests/methods/index.html), however. It seems to work fine with "Merge commit" and "Merge commit with semi-linear history". I would expect some issues with "Fast forward", but I have not tried.

Anyway, thanks for writing Git Town! I really love it!

Best, Manuel

kevgo reviewed (commented):
This PR is very well done. I really only have a few nitpicks. Sorry for the trouble with local testing and thanks for adding this important feature!

kevgo commented:
Please sync this branch to fix the CI build.

manuelprinz commented on features/sync/current_branch/feature_branch/features/pull_branch_strategy.feature:
```diff
@@ -28,11 +28,11 @@ Feature: with pull-branch-strategy set to "merge"
       | main    | local, origin | local main commit                                          |
       |         |               | origin main commit                                         |
-      |         |               | Merge remote-tracking branch 'origin/main'                 |
+      |         |               | Merge remote-tracking branch 'origin/main' into main       |
```
You are right, that box has 2.20.1. I'll revert the commit.

manuelprinz commented on src/hosting/gitlab.go:
```diff
@@ -46,14 +76,101 @@ func (d *GitlabDriver) NewPullRequestURL(branch, parentBranch string) (string, e
 	return fmt.Sprintf("%s/merge_requests/new?%s", d.RepositoryURL(), query.Encode()), nil
 }
 
+func (d *GitlabDriver) BaseURL() string {
+	return fmt.Sprintf("https://%s", d.hostname)
+}
```
Thanks!

manuelprinz commented on src/hosting/gitlab.go:
```diff
@@ -28,14 +34,38 @@ func NewGitlabDriver(config config) *GitlabDriver {
+		return result, err
+	}
+	if len(mergeRequests) != 1 {
+		return result, err
```
Added.

manuelprinz commented:
> This PR is very well done. I really only have a few nitpicks. Sorry for the trouble with local testing and thanks for adding this important feature!

Thank you for your kind words and quick reply!

kevgo commented:
Thanks, this is ready to go. We typically update code and documentation together (thanks to this being a mono-repo). Can you please mention this new GitLab support on https://github.com/git-town/git-town/blob/main/website/src/commands/ship.md and https://github.com/git-town/git-town/blob/main/website/src/quick-configuration.md#api-access-to-your-hosting-provider. Sorry that I didn't mention this earlier.

manuelprinz commented:
Sure, no problem! I was away for a couple of days and will update the MR during the next days.

manuelprinz commented:
I updated the documentation. Hope this is what you had in mind! I also put in some GitLab-specific doc because it might confuse GitLab users. I added it to the configuration section, as I did not find a better spot. Please let me know if I should change that!

kevgo commented:
Thanks for contributing this awesome feature! 🚀
